Aug 4 2021/Ch1ResnetV3.py

```diff
@@ -1,245 +1,225 @@
+            """
+            For loop for validating our model. For images and targets in our validation loader validate our model.
+            Similar to the training for loop used in our training function 
+            """
+            data, targets = data.to(device), targets.to(device)
+            # data = data.to(device) # send the images to the device either CPU or GPU
+            # targets.to(device) # send the labels to the device either CPU or GPU
+            outputs = model(data) # outputs = the model trained on our images
+            loss = criterion(outputs, targets) # loss = the loss function evaluated at the model trained on the images and the labels
+            valid_loss += loss.item() # Add this loss defined in the line above to the validation loss defined in line 147
+            _, predicted = torch.max(outputs.data,1) # returns the index of the highest prediction
+            # torch.max returns the maximum value of all elements in the input tensor
 
-"""
-2) track average training loss
-    - After training loop but before validation
-    - wanb.log({train_ave_loss':np.mean(train)
-"""
 
 
-"""
-3) track validation average loss
-    - loss_list is a list of all the loss values (i.e., loss_list.append(train_loss.item()) at each time step)
-    - at the end of the validation loop do
-    - wandb.log({‘val_ave_loss’: np.mean(val_loss_list)})
-"""
+        wandb.log({'Validation_average_loss': np.mean(valid_loss)}) # with weights and biases calculate the average validation loss and then log it
+
+
+def visualize_prediction(model,validation_loader,num_images=10): # Function to visualize predictions made with the model. This function takes an input our model, validation data and the number of images you want to display
+      model.eval() # Evaluate the model
+      images_so_far = 0 # images that have been shown so far
+      with torch.no_grad(): # Using torch
+         for i,(data,targets) in enumerate(validation_loader): # for image in data and targets located in the validation loader
+             data = data.to(device) # send the images to the device
+             targets = targets.to(device) # send the labels to the device
+             outputs = model(data) # run the images through the model
+             _,predicted = torch.max(outputs,1) # returns the index of the highest prediction
+             images_so_far +=1 # add one to the number of images used so far
+             image = data[i] # image = image tensor indexed at i
+             plt.imshow(image.cpu().squeeze(), cmap='gray') # plot the image
+             ground_truth = targets[i]
+             model_predicted = _[i]
+             title = ('ground truth',ground_truth,'model predicted',model_predicted) # set the title
+             plt.title(title) # add the title to the image
+             plt.show() # show the image
+             if images_so_far == num_images: # keep going over images until num_images = images_so_far
```

Comment:
can replace this logic with:

if i == num_images:
     return
     
     
no need to use the images_so_far tracker since the 'enumerate' iterator already does that